### ISSUE TITLE
ContinuousCallback: affect! is not always called at the crossed condition

### ISSUE THREAD

**Issue by MartinOtter** [SciML/DifferentialEquations.jl#551]
A state event should be triggered by ContinousCallback when the condition function **crosses** zero.  
Assume `z` is the condition and `z` is initially > 0. At the time instant of the event the callback function affect! or affect_neg! is called and `z` should then be <= 0 in order that the condition crossed zero. Typically, the crossing iteration determines a time interval in which the crossing of the condition appears and the integrator steps to the right limit of this interval when the iteration termations. It is then guaranteed that the condition <= 0.

This seems to be not the current behaviour as the following model demonstrates:
```
module Test_bouncingBall
# Model of a bouncing ball. x[1] is the height h, and x[2] is the velocity

using DifferentialEquations, PyPlot

function f(du,u,p,t)
  du[1] = u[2]
  du[2] = -9.81
end

function condition(u,t,integrator) # Event when event_f(u,t) == 0
  u[1]    # Event when height crosses from positive to negative
end

function affect_neg(integrator)
  println("h = ", integrator.u[1])
  integrator.u[2] = -0.8*integrator.u[2]
end

cb = ContinuousCallback(condition,nothing,affect_neg! = affect_neg)

u0 = [1.0,0.0]
tspan = (0.0, 3.0)
prob = ODEProblem(f,u0,tspan)
sol = solve(prob,Tsit5(),saveat=0.01,callback=cb)

figure(1)
clf()
plot(sol.t, getindex.(sol.u, 1), label="h")

end
```
This results in the following printout:
```
h = -2.338425295331619e-15
h = 1.4654949914765007e-14
h = -1.1669369619170325e-15
h = 7.706720075406634e-16
h = 2.606652068915516e-16
h = -1.6258013994107608e-15
```
This means that the integrator is not always at a time point after the zero crossing took place, but sometimes slightly before this point.

A temporary fix is to add a hysteresis to the condition:
```
function condition(u,t,integrator) # Event when event_f(u,t) == 0
  u[1]  + 1.e-12
end
```
resulting in
```
h = -9.997495282844048e-13
h = -9.839840025091685e-13
h = -9.943417128374115e-13
h = -9.977495486502297e-13
h = -1.0001823571158618e-12
h = -1.0000668227622088e-12
```

**Comment by ChrisRackauckas** [comment]
@kanav99 let's track this down.

**Comment by kanav99** [comment]
Yeah, I will get it to quick after DAE thing ends

**Comment by ChrisRackauckas** [comment]
Thanks for the report. This now passes and we've turned it into a test case for this fact.

**Comment by MartinOtter** [comment]
I have tested this again. I recognized that the crossing function `h` is now always >0.0, so it never crosses zero (if it started above zero). This is fine and is advantageous in cases where the crossing function defines a border that should not be crossed (as in the "bouncing ball" example above).

It would be helpful to **mention this behaviour in the documentation**, since traditionally (e.g. in DASSL, IDA) an integrator always guarantees that the zero-crossing function is either zero or is on the opposite side when the event occurs.

**Comment by ChrisRackauckas** [comment]
Yes, we should document this. Also, we're swapping out the nonlinear solver to make this more robust soon.

https://github.com/JuliaComputing/NonlinearSolve.jl

**Comment by MartinOtter** [comment]
I have experimented more with this feature, and I would like to advice that DifferentialEquations switches to the standard way to define zero crossing, or at least provides an option to use it (so the found root is bracketed in a time interval [t_left, t_right] and thet_ right value is reported as root, and not the t_left value, as it seems to be the case in DifferentialEquations: so the found root is either identical to zero or has an opposite sign as before the event occurs) . Reasons are:

### No longer safe to detect zero crossing for switching conditions

Assume that the discontinuity of the following equation:

```
y = u > 0 ? +1 : -1
```

should trigger a state event, so whenever u crosses zero, y should be +1 or -1 depending on the sign of u (using appropriate condition/affect functions) but the switching between the two if-branches should be done at an event instant and not during continuous integration. Assume that u is positive and becomes negative. So, initially, u>0, y=+1. With the standard root handling an event is triggered, when u <= 0. At the event, the if-clause is switched to the else-branch and at this time instant everything is consistent again (u is exactly zero or less than zero and therefore y=-1).

With the DifferentialEquations approach, an event is triggered when u > 0. However, it is then not possible to switch the if-clause at the event instant, because otherwise the relation (u > 0) and the selected if-branch are not consistent to each other (u>0, but the "else" branch shall be selected). The only way to fix this is to cheat and use as crossing function z not "u" but "z = u-eps", in order that at an event instant z>0, but u<0.  Now, the question is how large the "eps" has to be selected. This depends on the (unknown) epsilon strategy of DifferentialEquations **and** on the **unknown** function u(t). Therefore, it is **impossible** to implement this discontinuity in a safe way.

### No longer safe to handle zeno-type systems

Simulate the bouncing ball example for a longer time (say stopTime=5 s).  You will see that the ball is then no longer bouncing but flying through the ground:

![image](https://user-images.githubusercontent.com/7248038/96493973-fc4ed400-1245-11eb-90a1-9e7d0e16b1f2.png)

The reason is that DifferentialEquations no longer triggers an event, when the event is too close to the previous event. With the standard root function technique, an event is **always** triggered when a zero crossing occurs. The user is then responsible to provide an appropriate action. For the bouncing ball example a completely safe strategy is to monitor the derivative of the height: If the rebound velocity is too small or if an event is triggered where the velocity changes sign from positive to negative and at this event instant the height of the ball is zero or negative, then switch to the model "ball is on ground" (in the simplest case by changing the differential equation and setting the acceleration of the ball to zero).

With the zero-crossing technique of DifferentialEquations this approach cannot be used, because the critical situation (no event anymore) is not reported to the user. Again, the only way is to introduce an epsilon for the zero crossing (so trigger an event if height-eps crosses zero), but as before, it is not possible to select in a safe way the epsilon.

**To summarize, please support the standard root function technique where the zero crossing function has an opposite sign after the event with respect before the event (or be identically zero)**

**Comment by ChrisRackauckas** [comment]
Yes, we know about this issue with Roots.jl, which is why we're going to swap the

**Comment by MartinOtter** [comment]
Any progress here, that it is possible to get the right limit at a zero crossing (this missing feature makes really problems in our applications and requires to use unnecessary hysteresis that may not work well)?

**Comment by ChrisRackauckas** [comment]
I'm not sure there's anything left here?

https://diffeq.sciml.ai/stable/features/callback_functions/#ContinuousCallback

> rootfind=LeftRootFind: This is a flag to specify the type of rootfinding to do for finding event location. If this is set to LeftRootfind, the solution will be backtracked to the point where condition==0 and if the solution isn't exact, the left limit of root is used. If set to RightRootFind, the solution would be set to the right limit of the root. Otherwise the systems and the affect! will occur at t+dt.

and

https://diffeq.sciml.ai/stable/features/callback_functions/#Handling-Accumulation-Points

> Handling Accumulation Points. Now let's take a look at the bouncing ball with friction. After the bounce, we will send the velocity to -v/2. Since the velocity is halving each time, we should have Zeno-like behavior and see an accumulation point of bounces. We will use some extra parameters to count the number of bounces (to infinity) and find the accumulation point. Let's watch!

That example works for that case.

**Comment by MartinOtter** [comment]
I am using DifferentialEquations 6.21.0, but here the options LeftRootFind or RightRootFind seem to be not available (or I make a mistake).

**Comment by ChrisRackauckas** [comment]
They are available. Are you not namespacing it appropriately? `DiffEqBase.LeftRootFind`? The keyword argument is there, and it's been there for quite awhile.

**Comment by MartinOtter** [comment]
The options are in SciMLBase, not in DiffEqBase (I searched with the editor in the downloaded packages):
```
import DifferentialEquations

...., rootfind = DifferentialEquations.SciMLBase.RightRootFind
```
Thanks very much. This works and the issues in my application are gone.

**Comment by ChrisRackauckas** [comment]
Cool, good to hear. And I improved the docstring to mention the SciMLBase namespacing, so this should all be cleared up. Cheers.
